New line c preprocesssor

C preprocessor takes first token of the next line as part of current line:

```
<SNIP>/raw_source.cpp:2:19: Included file:
comas.h:1:1: Error: 1. Unknown preprocessor directive
#define COMBINE_(x, y) x y
^
```

To reproduce:

add two include files on top of addVectors.okl (example 01 cpp case):

addVectors.okl:
-----------------

```cpp
#include "comas.h"
#include "macros.h"

// 1- Note no new line between two includes above.
// 2- To make code work put a new line between two above includes.

@kernel void addVectors(const int entries,
                        const float *a,
                        const float *b,
                        float *ab) {
  //superclever(ab);
  callme(&ab[0], &ab[1] ? &ab[1] : &ab[0] , &ab[2] ? &ab[2] : &ab[1], &ab[3] ? &ab[3] : &ab[2] );
  for (int i = 0; i < entries; ++i; @tile(16, @outer, @inner)) {
    ab[i] = a[i] + b[i];
  }
}
```

comas.h:
----------

```cpp
#define COMBINE_(x, y) x y
inline void callme(float *a, float *b, float *c, float *d){ *a;}
```

macros.h:
-----------

```cpp
#define superclever(z) COMBINE_(callme, (&z[0], &z[1] ? &z[1] : &z[0] , &z[2] ? &z[2] : &z[1], &z[3] ? &z[3] : &z[2] ))
```

--- a/include/occa/lang/preprocessor.hpp
+++ b/include/occa/lang/preprocessor.hpp
@@ -37,8 +37,7 @@ namespace occa {
       extern const int finishedIf;
     }
 
-    class preprocessor_t : public withInputCache<token_t*, token_t*>,
-                           public withOutputCache<token_t*, token_t*> {
+    class preprocessor_t : public withCache<token_t*, token_t*> {
     public:
       typedef void (preprocessor_t::*processDirective_t)(identifierToken &directive);
       typedef std::map<std::string, processDirective_t> directiveMap;
@@ -71,7 +70,6 @@ namespace occa {
 
       //---[ Misc ]---------------------
       tokenizer_t *tokenizer;
-      bool hasLoadedTokenizer;
 
       strVector includePaths;
       //================================

--- a/include/occa/lang/processingStages.hpp
+++ b/include/occa/lang/processingStages.hpp
@@ -11,6 +11,7 @@ namespace occa {
     typedef streamMap<token_t*, token_t*>       tokenMap;
     typedef withInputCache<token_t*, token_t*>  tokenInputCacheMap;
     typedef withOutputCache<token_t*, token_t*> tokenOutputCacheMap;
+    typedef withCache<token_t*, token_t*>       tokenCacheMap;
 
     class newlineTokenFilter : public tokenFilter {
     public:
@@ -29,8 +30,7 @@ namespace occa {
       virtual void fetchNext();
     };
 
-    class externTokenMerger : public tokenInputCacheMap,
-                              public tokenOutputCacheMap {
+    class externTokenMerger : public tokenCacheMap {
     public:
       externTokenMerger();
       externTokenMerger(const externTokenMerger &other);

--- a/include/occa/lang/stream.hpp
+++ b/include/occa/lang/stream.hpp
@@ -28,6 +28,7 @@ namespace occa {
 
       virtual void setNext(output_t &out) = 0;
 
+      virtual void clearCache();
       virtual void* passMessageToInput(const occa::properties &props);
       void* getInput(const std::string &name);
 
@@ -62,6 +63,7 @@ namespace occa {
 
       stream& clone() const;
 
+      void clearCache();
       void* passMessageToInput(const occa::properties &props);
       void* getInput(const std::string &name);
 
@@ -93,6 +95,7 @@ namespace occa {
       virtual streamMap& clone() const;
       virtual streamMap& clone_() const = 0;
 
+      virtual void clearCache();
       virtual void* passMessageToInput(const occa::properties &props);
     };
 
@@ -125,6 +128,8 @@ namespace occa {
 
       virtual void setNext(input_t &out);
       virtual bool isValid(const input_t &value) = 0;
+
+      virtual void clearCache();
     };
 
     template <class input_t>
@@ -155,6 +160,8 @@ namespace occa {
       void pushInput(const input_t &value);
 
       void getNextInput(input_t &value);
+
+      virtual void clearCache();
     };
 
     template <class input_t, class output_t>
@@ -172,6 +179,18 @@ namespace occa {
       void pushOutput(const output_t &value);
 
       virtual void fetchNext() = 0;
+
+      virtual void clearCache();
+    };
+
+    template <class input_t, class output_t>
+    class withCache : public withInputCache<input_t, input_t>,
+                      public withOutputCache<output_t, output_t> {
+    public:
+      withCache();
+      withCache(const withCache<input_t, output_t> &other);
+
+      virtual void clearCache();
     };
     //====================================
   }

--- a/include/occa/lang/stream.tpp
+++ b/include/occa/lang/stream.tpp
@@ -5,6 +5,9 @@ namespace occa {
     baseStream<output_t>::~baseStream() {}
 
     template <class output_t>
+    void baseStream<output_t>::clearCache() {}
+
+    template <class output_t>
     void* baseStream<output_t>::passMessageToInput(const occa::properties &props) {
       return NULL;
     }
@@ -62,6 +65,11 @@ namespace occa {
         return *(new stream());
       }
       return *(new stream(&(head->clone())));
+    }
+
+    template <class output_t>
+    void stream<output_t>::clearCache() {
+      head->clearCache();
     }
 
     template <class output_t>
@@ -138,6 +146,13 @@ namespace occa {
     }
 
     template <class input_t, class output_t>
+    void streamMap<input_t, output_t>::clearCache() {
+      if (input) {
+        input->clearCache();
+      }
+    }
+
+    template <class input_t, class output_t>
     void* streamMap<input_t, output_t>::passMessageToInput(const occa::properties &props) {
       if (input) {
         return input->passMessageToInput(props);
@@ -199,6 +214,13 @@ namespace occa {
     }
 
     template <class input_t>
+    void streamFilter<input_t>::clearCache() {
+      streamMap<input_t, input_t>::clearCache();
+      usedLastValue = true;
+      isEmpty_ = true;
+    }
+
+    template <class input_t>
     streamFilterFunc<input_t>::streamFilterFunc(bool (*func_)(const input_t &value)) :
       func(func_) {}
 
@@ -249,6 +271,12 @@ namespace occa {
     }
 
     template <class input_t, class output_t>
+    void withInputCache<input_t, output_t>::clearCache() {
+      streamMap<input_t, output_t>::clearCache();
+      inputCache.clear();
+    }
+
+    template <class input_t, class output_t>
     withOutputCache<input_t, output_t>::withOutputCache() :
       outputCache() {}
 
@@ -278,6 +306,28 @@ namespace occa {
     template <class input_t, class output_t>
     void withOutputCache<input_t, output_t>::pushOutput(const output_t &value) {
       outputCache.push_back(value);
+    }
+
+    template <class input_t, class output_t>
+    void withOutputCache<input_t, output_t>::clearCache() {
+      streamMap<input_t, output_t>::clearCache();
+      outputCache.clear();
+    }
+
+    template <class input_t, class output_t>
+    withCache<input_t, output_t>::withCache() {}
+
+    template <class input_t, class output_t>
+    withCache<input_t, output_t>::withCache(
+      const withCache<input_t, output_t> &other
+    ) :
+        withInputCache<input_t, input_t>(other),
+        withOutputCache<output_t, output_t>(other) {}
+
+    template <class input_t, class output_t>
+    void withCache<input_t, output_t>::clearCache() {
+      withInputCache<input_t, input_t>::clearCache();
+      withOutputCache<output_t, output_t>::clearCache();
     }
   //====================================
   }

--- a/src/lang/parser.cpp
+++ b/src/lang/parser.cpp
@@ -223,6 +223,7 @@ namespace occa {
     void parser_t::setSource(const std::string &source,
                              const bool isFile) {
       clear();
+      stream.clearCache();
 
       if (isFile) {
         tokenizer.set(new file_t(source));

--- a/src/lang/preprocessor.cpp
+++ b/src/lang/preprocessor.cpp
@@ -52,8 +52,7 @@ namespace occa {
     }
 
     preprocessor_t::preprocessor_t(const preprocessor_t &pp) :
-      withInputCache(pp),
-      withOutputCache(pp) {
+      withCache(pp) {
       *this = pp;
     }
 
@@ -117,7 +116,6 @@ namespace occa {
       errors   = 0;
 
       tokenizer = NULL;
-      hasLoadedTokenizer = false;
     }
 
     void preprocessor_t::clear() {
@@ -130,7 +128,6 @@ namespace occa {
       warnings = 0;
 
       tokenizer = NULL;
-      hasLoadedTokenizer = false;
 
       while (inputCache.size()) {
         delete inputCache.front();
@@ -486,9 +483,8 @@ namespace occa {
     //==================================
 
     void preprocessor_t::loadTokenizer() {
-      if (!hasLoadedTokenizer) {
+      if (!tokenizer) {
         tokenizer = (tokenizer_t*) getInput("tokenizer_t");
-        hasLoadedTokenizer = true;
       }
     }
 
@@ -1371,6 +1367,7 @@ namespace occa {
       }
 
       // Push source after updating origin to the [\n] token
+      input->clearCache();
       tokenizer->pushSource(header);
     }
 

--- a/src/lang/processingStages.cpp
+++ b/src/lang/processingStages.cpp
@@ -83,8 +83,7 @@ namespace occa {
     externTokenMerger::externTokenMerger() {}
 
     externTokenMerger::externTokenMerger(const externTokenMerger &other) :
-      tokenInputCacheMap(other),
-      tokenOutputCacheMap(other) {}
+      tokenCacheMap(other) {}
 
     tokenMap& externTokenMerger::clone_() const {
       return *(new externTokenMerger(*this));

